libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -68,818 +66,405 @@ static const uint16_t ship_feature_pruned_log = 1;
+         fc::raw::pack(appender, num_blocks_in_log);
       }
    }
 
-   fc::cfile& get_entry(uint32_t block_num, state_history_log_header& header) {
-      EOS_ASSERT(block_num >= _begin_block && block_num < _end_block, chain::plugin_exception,
-                 "read non-existing block in ${name}.log", ("name", name));
-      log.seek(get_pos(block_num));
-      read_header(header);
-      return log;
-   }
-
-   std::optional<chain::block_id_type> get_block_id_i(uint32_t block_num) {
-      auto result = catalog.id_for_block(block_num);
-      if (!result) {
-         if (block_num >= _begin_block && block_num < _end_block) {
-            state_history_log_header header;
-            get_entry(block_num, header);
-            EOS_ASSERT(chain::block_header::num_from_id(header.block_id) == block_num, chain::plugin_exception,
-                       "header id does not match requested ${a} != ${b}", ("a", chain::block_header::num_from_id(header.block_id))("b", block_num));
-            return header.block_id;
-         }
-         return {};
-      }
-      EOS_ASSERT(chain::block_header::num_from_id(*result) == block_num, chain::plugin_exception,
-                 "catalog id does not match requested ${a} != ${b}", ("a", chain::block_header::num_from_id(*result))("b", block_num));
-      return result;
-   }
-
-   //file position must be at start of last block's suffix (back pointer)
-   //called from open_log / ctor
-   bool get_last_block() {
-      state_history_log_header header;
-      uint64_t                 suffix;
-
-      fc::raw::unpack(log, suffix);
-      const size_t after_suffix_pos = log.tellp();
-      if (suffix > after_suffix_pos || suffix + state_history_log_header_serial_size > after_suffix_pos) {
-         elog("corrupt ${name}.log (2)", ("name", name));
-         return false;
-      }
-      log.seek(suffix);
-      read_header(header, false);
-      if (!is_ship(header.magic) || !is_ship_supported_version(header.magic) ||
-          suffix + state_history_log_header_serial_size + header.payload_size + sizeof(suffix) != after_suffix_pos) {
-         elog("corrupt ${name}.log (3)", ("name", name));
-         return false;
-      }
-      _end_block    = chain::block_header::num_from_id(header.block_id) + 1;
-      last_block_id = header.block_id;
-      if (_begin_block >= _end_block) {
-         elog("corrupt ${name}.log (4)", ("name", name));
-         return false;
-      }
-      return true;
+   std::optional<chain::block_id_type> get_block_id(uint32_t block_num) {
+      if(block_num >= _begin_block && block_num < _end_block)
+         return log.unpack_from<log_header>(get_pos(block_num)).block_id;
+      return std::nullopt;
    }
 
-   void prune(const fc::log_level& loglevel) {
-      auto prune_config = std::get_if<state_history::prune_config>(&_config);
-
+ private:
+   void prune() {
       if(!prune_config)
          return;
       if(_end_block - _begin_block <= prune_config->prune_blocks)
```

Comment:
This does have an off-by-one feel to it but in the other direction?

Let's say `begin=10`, `end=21`, that means we have 10 blocks total, and really we should only allow pruning 9 of those because we need a single valid block at the end of the log otherwise the log has no valid structure.

Now if `prune_blocks=10` that's `if(21-10<=10)` so `if(11<=10)`, this results in **false** so we _do not return_ ergo we allow pruning. 

Even if we ignore the "keep 1 entry" rule from above, if `prune_blocks=11` that's `if(21-10<=11)` so `if(11<=11)`, this also results in **false** so we _do not return_ ergo we allow pruning here too.

Since this is not new to this refactor I'd suggest we defer this to a separate issue (it might even need to be fixed in 5.0). Added as #272